In [1]:
import tensorflow as tf
import numpy as np
import json
import pandas as pd
import nltk
from tensorflow.keras.preprocessing.text import Tokenizer
from tensorflow.keras.layers import Input,Embedding,LSTM,Dense,GlobalMaxPooling1D,Flatten
from tensorflow.keras.models import Model
import matplotlib.pyplot as plt
import speech_recognition as sr
import pyttsx3
import subprocess as s
import datetime
from datetime import time
import os
import sqlite3
import nltk
import string
# nltk.download('stopwords')
nltk.download('punkt')
from nltk.tokenize import WordPunctTokenizer
from nltk.tokenize import word_tokenize
from nltk.corpus import stopwords

from nltk.tokenize import WordPunctTokenizer
from nltk.tokenize import word_tokenize
from nltk.corpus import stopwords


[nltk_data] Downloading package punkt to
[nltk_data]     C:\Users\DELL\AppData\Roaming\nltk_data...
[nltk_data]   Package punkt is already up-to-date!


In [2]:
import speech_recognition as sr
import pyttsx3
import os
import datetime

In [3]:
# nltk.download('stopwords')
# nltk.download('punkt')

stop_words={'but', 'aren', 'by', 'during', "isn't", "won't", 'through', 'only', 'up', 'we', 'o', 'very', 'off', 'here', 'hasn', 
            'does', 'that', "shouldn't", 'wasn', 'of', 'shan', 'yourselves', 'my', 'd', "you'd", 'further', 'too', "that'll", 
            'these', 'or', 'before', 'each', 'm', 'all', 'at', "didn't", "doesn't", 'be', 'the', 'are', 'been', 'being', 'didn'
            , 'if', 'their', 'until', 'i', 'hadn', 'against', 'to', 'once', 'your', 'am', 'just', 'again', 'with', 'while', 
            'between', "needn't", 'a', "it's", 'where', 'themselves', 'after', 'don', 'ma', 'when', "mustn't", 'ours', 'can',
            "hasn't", 'both', 'is', 'under', 'have', 't', 'he', 'she', 'will', 'hers', 'were', 'own', "mightn't", 
            'doing', 'shouldn', 'herself', 'wouldn', 'above', 'yours', 've', "don't", 'an', 'any', 'same', 'no', 'do',
            'there', 'nor', 'some', "weren't", 'our', 'll', 'did', 'should', "aren't", 'won', 're', 'not', 'this', 'than', 
            's', 'has', 'had', "should've", 'what', "couldn't", 'then', 'haven', 'into', 'from', "shan't", 'y', 'such',
            'below', 'few', 'they', 'having', 'her', "wasn't", 'was', 'needn', 'ourselves', "you've", "wouldn't", 
            'on', 'over', 'because', 'as', "haven't", 'its', 'you', 'isn', "she's", 'those', "hadn't", 'more', 'myself',
            'him', 'which', 'it', 'weren', "you'll", 'for', 'himself', 'and', 'itself', 'now', 'doesn', 'other', 'me', 'them',
            'out', 'ain', "you're", 'most', 'could not', 'his', 'mightn', 'who', 'in', 'so', 'whom', 'yourself', 'why', 'theirs',
            'mustn', 'down',"tell"}

In [4]:
with open("content.json") as content:
    data1=json.load(content)

In [5]:
def words_to_list(s):
    a=[]
    ns=""
    s=s+" "
    for i in range(len(s)):
        if s[i]==" ":
            a.append(ns)
            ns=""
        else:
            ns=ns+s[i]
    a=list(set(a))
    return a

In [6]:
tags=[]

responses={}
patt=[]
filter_sentence = []
for intent in data1['intents']:
    responses[intent['tag']]=intent['responses']
    for lines in intent['patterns']:
        patt.append(lines)
        tags.append(intent['tag'])

# # stop_words = set(stopwords.words('english'))
# filter_sentence = [w for w in ntokens if not w.lower() in stop_words]
# for ntokens in patterns:
#     word_tokens = word_tokenize(ntokens)
#     filtered_sentence = [w for w in word_tokens if not w.lower() in stop_words]
#     #with no lower case conversion
#     filtered_sentence = []
#     for w in word_tokens:
#         if w not in stop_words:
#             filtered_sentence.append(w)
#             patt.append(str(" ".join([c for c in filter_sentence])))
    
# patt
# patt
len(tags)

43

In [7]:
# tags=[]
# patterns=[]
# responses=[]
# for intent in data1['intents']:
# #      responses[intent['tag']]=intent['responses']
#     for res in intent['responses']:
#         for lines in intent['patterns']:
#             patterns.append(lines)
#             responses.append(res)
#             tags.append(intent['tag'])
        
patterns=[]
for i in patt:  
    word_tokens = word_tokenize(i)
    # converts the words in word_tokens to lower case and then checks whether 
    #they are present in stop_words or not
    filtered_sentence = [w for w in word_tokens if not w.lower() in stop_words]
    #with no lower case conversion
    filtered_sentence = []

    for w in word_tokens:
        if w not in stop_words:
            filtered_sentence.append(w)
    patterns.append(" ".join(filtered_sentence))
    

# patterns

In [8]:
data=pd.DataFrame({"inputs":patterns,
                   "tags":tags
                  })

In [9]:
import string 
data['inputs']=data['inputs'].apply(lambda wrd:[ltrs.lower() for ltrs in wrd if ltrs not in string.punctuation])
data['inputs']=data['inputs'].apply(lambda wrd:''.join(wrd))
data
data

inputs                     tags
0                                 college name             college name
1                                about college             college name
2                                 name college             college name
3                           idea about college             college name
4                       deparment name college          department name
5                          stream name college          department name
6                     availabe courses college          department name
7                  how many department college          department name
8                        count department name          department name
9                         available department          department name
10       how many department available college          department name
11                           number department          department name
12                      count total department          department name
13                    total department college          department name
14                      vision mission college       vision and mission
15                      vision mission college       vision and mission
16                               goals college       vision and mission
17                              vision college       vision and mission
18                             mission college       vision and mission
19                    established year college                     year
20                             college started                     year
21                        year college started                     year
22                    year college extablished                     year
23                      principal name college                principal
24                             about principal                principal
25                 information about principal                principal
26                              principal name                principal
27                         library information                  library
28                   information about library                  library
29                         information library                  library
30                   snjb college library good                  library
31                  how infrastructure library                  library
32             placement information 2021 2022                Placement
33                             about placement                Placement
34              college provide good recruters                Placement
35                         name vice principal           vice principal
36                        about vice principal           vice principal
37            information about vice principal           vice principal
38           how many labs computer department  computer department lab
39             number labs computer department  computer department lab
40                count computer depatment lab  computer department lab
41              total labs computer department  computer department lab
42  how many labs avilable computer department  computer department lab

In [10]:
try:
    value = pd.DataFrame(data.loc[data['inputs'] == 'library information'])
    v=value.iloc[0]['tags']
    print(v)
except:
    if len(value)<1:
        print("None")
   

library


In [11]:
def check(text):
    try:
        value = pd.DataFrame(data.loc[data['inputs'] == text])
        v=value.iloc[0]['tags']
#         print(v)
    except:
        if len(value)<1:
            talk("sorry we i didn't find please ask again")
            ask()
#     print(value[0])
#     res = data.loc[data['inputs'] == text]['tags']
#     print(value[0:1])
#     print(responses[value[0:1]])

    return responses[v]

In [12]:
def chat(txt_pun):
    ntokens = word_tokenize(txt_pun)
    # stop_words = set(stopwords.words('english'))
    filter_sentence = [w for w in ntokens if not w.lower() in stop_words]
    filter_sentence = []
    for w in ntokens:
        if w not in stop_words:
            filter_sentence.append(w)

#     print(ntokens, "\n")
#     print(filter_sentence, "\n")
    punc = string.punctuation
            # print(punc)
    txt_pun = str(" ".join([c for c in filter_sentence if c not in string.punctuation]))
#     print(txt_pun)
    ans=check(txt_pun)
    talk(ans)
    #         record=mycol.find({"pattern": txt_pun})
    #         print(record)

In [13]:
def ask():
    while True:
        q=input('')
        if q=="stop" or q=="STOP" or q=="Stop" or q=="quit":
            print("Thank you, have a nice day:)")
            break
        chat(q)

In [ ]:
ask()

# Voice Assistant


In [15]:
listener = sr.Recognizer()
engine = pyttsx3.init()
voices = engine.getProperty('voices')
engine.setProperty('rate', 170)
engine.setProperty('volume', 0.5)
engine.setProperty('voice', 'english+f5')

def username():
    talk("What should i call you")
    uname = take_command()
    talk("Welcome")
    talk(uname)
    talk("How can i Help you")

def talk(text):
    engine.say(text)
    engine.runAndWait()


def take_command():
    try:
        with sr.Microphone() as source:
            listener.pause_threshold = 1
            print("listening...")
            voice = listener.listen(source, timeout=3, phrase_time_limit=5)
            r = sr.Recognizer()
            r.energy_threshold = 300
            r.adjust_for_ambient_noise(source, 1.2)
            command = listener.recognize_google(voice, language='en-in')
            command = command.lower()
            if 'snjbya' in command:
                command = command.replace('snjbya', '')

    except:
        print("error")
        print("Unable to Recognize your voice.")
        return "None"

    return command


def wishMe():
    hour = int(datetime.datetime.now().hour)
    if hour >= 0 and hour < 12:
        talk("Good Morning!")

    elif hour >= 12 and hour < 18:
        talk("Good Afternoon!")

    else:
        talk("Good Evening!")

    assname = ("SNJBEEA")
    talk("I am your Voice Assistant SNJBEEA")

def ask():
    while True:
        command = take_command().lower()
#         print(command.result[0][0].transcript)
        if command=="stop":
            talk("Thank you for asking")
            break
        chat(command)
    

text = ""

if __name__ == '__main__':
    clear = lambda: os.system('cls')
    clear()
    wishMe()
    username()
    ask()

listening...
result2:
{   'alternative': [   {'confidence': 0.73494297, 'transcript': 'Pooja'},
                       {'transcript': 'Puja'},
                       {'transcript': 'Pooja Bahu Ka Pratibha'},
                       {'transcript': 'Puja Bahu Ka Pratibha'}],
    'final': True}
listening...
result2:
{   'alternative': [   {   'confidence': 0.88687539,
                           'transcript': 'tell me college name'}],
    'final': True}
listening...
result2:
{   'alternative': [   {   'confidence': 0.87642086,
                           'transcript': 'total department in smjb'},
                       {'transcript': 'total department in smjp'},
                       {'transcript': 'total department in smj.b'},
                       {'transcript': 'total department in snjb'},
                       {'transcript': 'total department in smj.v'}],
    'final': True}
listening...
result2:
{   'alternative': [   {   'confidence': 0.88687533,
                           'transcrip

UnboundLocalError: local variable 'v' referenced before assignment

listening...
result2:
{   'alternative': [{'confidence': 0.88687539, 'transcript': 'Pratibha'}],
    'final': True}
listening...
result2:
{   'alternative': [   {   'confidence': 0.88687539,
                           'transcript': 'tell me college name'}],
    'final': True}
college name
listening...
error
Unable to Recognize your voice.
listening...
result2:
{   'alternative': [   {   'confidence': 0.88687533,
                           'transcript': 'tell me something about department'}],
    'final': True}
listening...
result2:
{   'alternative': [   {   'confidence': 0.80620319,
                           'transcript': 'about available departments'},
                       {'transcript': 'available departments'},
                       {'transcript': 'about available department'},
                       {'transcript': 'available department'},
                       {'transcript': 'about available'}],
    'final': True}
listening...
result2:
{   'alternative': [   {   'confidence': 0.88687539,
                           'transcript': 'library information'}],
    'final': True}
library
listening...
result2:
[]
error
Unable to Recognize your voice.
listening...
result2:
{   'alternative': [   {   'confidence': 0.79089665,
                           'transcript': 'tablet working mahiti'},
                       {'transcript': 'tablet working'},
                       {'transcript': 'working mahiti'},
                       {'transcript': 'working'}],
    'final': True}
listening...
result2:
{   'alternative': [   {   'confidence': 0.79144037,
                           'transcript': 'about availability department '
                                         'available department'},
                       {   'transcript': 'availability department available '
                                         'department'},
                       {   'transcript': 'available department available '
                                         'department'},
                       {'transcript': 'availability'},
                       {'transcript': 'availability department'}],
    'final': True}
listening...
result2:
{   'alternative': [   {   'confidence': 0.85622609,
                           'transcript': 'confidence courier your decide '
                                         'precisely'},
                       {   'transcript': 'confidence courier your decide '
                                         'decide leave'},
                       {   'transcript': 'confidence courier your decide '
                                         'decides leave'},
                       {   'transcript': 'confidence courier your decide '
                                         'preciously'},
                       {   'transcript': 'confidence courier your decide '
                                         'preciously your'}],
    'final': True}
listening...
result2:
{   'alternative': [   {   'confidence': 0.78409123,
                           'transcript': 'Itni result dilana'},
                       {'transcript': 'result dilana'},
                       {'transcript': 'Itni result Dila na'},
                       {'transcript': 'result Dila na'}],
    'final': True}
listening...
result2:
{   'alternative': [   {'confidence': 0.88687539, 'transcript': 'energy under'},
                       {'transcript': 'energy underscore'},
                       {'transcript': 'energy underscore threshold'},
                       {'transcript': 'energy underscore 3'},
                       {'transcript': 'energy underscore threshold hacked'}],
    'final': True}
listening...
result2:
[]
error
Unable to Recognize your voice.
listening...
error
Unable to Recognize your voice.

In [11]:
import string 
data['inputs']=data['inputs'].apply(lambda wrd:[ltrs.lower() for ltrs in wrd if ltrs not in string.punctuation])
data['inputs']=data['inputs'].apply(lambda wrd:''.join(wrd))
data

inputs                     tags
0                        college name             college name
1                       about college             college name
2                        name college             college name
3             give idea about college             college name
4                   department s name          department name
5                         stream name          department name
6                          field name          department name
7         how many department college          department name
8               count department name          department name
9                available department          department name
10         provide department details          department name
11                            courses          department name
12           which courses available           department name
13                              count  department count,number
14                        stream name  department count,number
15                         field name  department count,number
16        how many department college  department count,number
17              count department name  department count,number
18               available department  department count,number
19                number departments   department count,number
20                how many department  department count,number
21                   how many courses  department count,number
22             vision mission college       vision and mission
23             vision mission college       vision and mission
24                      goals college       vision and mission
25           established year college                     year
26                    college started                     year
27               year college founded                     year
28                    college founded                     year
29             principal name college                principal
30                    about principal                principal
31                library information                  library
32          information about library                  library
33                information library                  library
34                 describe libraries                  library
35              details about library                  library
36    placement information 2021 2022                Placement
37                    about placement                Placement
38                name vice principal           vice principal
39               about vice principal           vice principal
40         information vice principal           vice principal
41       datails about vice principal           vice principal
42  how many labs computer department  computer department lab
43    number labs computer department  computer department lab
44       count computer depatment lab  computer department lab

In [26]:
from tensorflow.keras.preprocessing.text import Tokenizer
tokenizer=Tokenizer(num_words=2000)
tokenizer.fit_on_texts(data['inputs'])
train=tokenizer.texts_to_sequences(data['inputs'])
from tensorflow.keras.preprocessing.sequence import pad_sequences
x_train=pad_sequences(train)
from sklearn.preprocessing import LabelEncoder
le=LabelEncoder()
y_train=le.fit_transform(data['tags'])
def check(text):
    value = pd.DataFrame(data.loc[data['inputs'] == text])
    v=value.iloc[0]['tags']
    if responses[v]=="":
        return "Sorry"
#     print(value[0])
#     res = data.loc[data['inputs'] == text]['tags']
#     print(value[0:1])
#     print(responses[value[0:1]])

    return responses[v]

In [13]:
input_shape=x_train.shape[1]
print(input_shape)

5


In [14]:
vocabulary=len(tokenizer.word_index)
print("No Of Unique Words:",vocabulary)
output_len=le.classes_.shape[0]
print("Output length:",output_len)

No Of Unique Words: 40
Output length: 10


In [15]:
i=Input(shape=(input_shape,))
x=Embedding(vocabulary+1,10)(i)
x=LSTM(10,return_sequences=True)(x)
x=Flatten()(x)
x=Dense(output_len,activation="softmax")(x)
model=Model(i,x)

In [16]:
model.compile(loss="sparse_categorical_crossentropy",optimizer='adam',metrics=['accuracy'])

In [17]:
train=model.fit(x_train,y_train,epochs=200)

Epoch 1/200
2/2 [==============================] - 2s 8ms/step - loss: 2.2969 - accuracy: 0.3333
Epoch 2/200
2/2 [==============================] - 0s 4ms/step - loss: 2.2929 - accuracy: 0.2222
Epoch 3/200
2/2 [==============================] - 0s 8ms/step - loss: 2.2893 - accuracy: 0.2222
Epoch 4/200
2/2 [==============================] - 0s 4ms/step - loss: 2.2853 - accuracy: 0.2000
Epoch 5/200
2/2 [==============================] - 0s 5ms/step - loss: 2.2815 - accuracy: 0.2000
Epoch 6/200
2/2 [==============================] - 0s 4ms/step - loss: 2.2770 - accuracy: 0.2000
Epoch 7/200
2/2 [==============================] - 0s 5ms/step - loss: 2.2730 - accuracy: 0.2000
Epoch 8/200
2/2 [==============================] - 0s 5ms/step - loss: 2.2683 - accuracy: 0.2000
Epoch 9/200
2/2 [==============================] - 0s 6ms/step - loss: 2.2636 - accuracy: 0.2000
Epoch 10/200
2/2 [==============================] - 0s 6ms/step - loss: 2.2581 - accuracy: 0.2000
Epoch 11/200
2/2 [===========

2/2 [==============================] - 0s 4ms/step - loss: 0.8748 - accuracy: 0.6889
Epoch 168/200
2/2 [==============================] - 0s 6ms/step - loss: 0.8699 - accuracy: 0.6889
Epoch 169/200
2/2 [==============================] - 0s 7ms/step - loss: 0.8659 - accuracy: 0.6889
Epoch 170/200
2/2 [==============================] - 0s 6ms/step - loss: 0.8613 - accuracy: 0.6889
Epoch 171/200
2/2 [==============================] - 0s 7ms/step - loss: 0.8568 - accuracy: 0.6889
Epoch 172/200
2/2 [==============================] - 0s 6ms/step - loss: 0.8520 - accuracy: 0.6889
Epoch 173/200
2/2 [==============================] - 0s 4ms/step - loss: 0.8473 - accuracy: 0.6889
Epoch 174/200
2/2 [==============================] - 0s 7ms/step - loss: 0.8425 - accuracy: 0.6889
Epoch 175/200
2/2 [==============================] - 0s 5ms/step - loss: 0.8377 - accuracy: 0.6889
Epoch 176/200
2/2 [==============================] - 0s 6ms/step - loss: 0.8332 - accuracy: 0.7111
Epoch 177/200
2/2 [=====

In [ ]:
import random
while True:
    texts_p=[]
    prediction_input=input('You:')
    prediction_input=[letters.lower() for letters in prediction_input if letters not in string.punctuation ]
    prediction_input=''.join(prediction_input)
    texts_p.append(prediction_input)
    print(prediction_input)

    prediction_input=tokenizer.texts_to_sequences(texts_p)
    print(prediction_input)
    prediction_input=np.array(prediction_input).reshape(-1)
    prediction_input=pad_sequences([prediction_input],input_shape)
    print(prediction_input)

    output=model.predict(prediction_input)
    output=output.argmax()
    print(output)

    response_tag=le.inverse_transform([output])[0]
    print("Going Marry:",random.choice(responses[response_tag]))

    if response_tag=="goodbye":
        break

You:details about vice principal
details about vice principal
[[17, 4, 11, 5]]
[[ 0 17  4 11  5]]
1/1 [==============================] - 0s 432ms/step
7
Going Marry: Sanghavi Sir is vice principle of SNJB's KBJ College of Engineering Chandwad
You:name of vice principal
name of vice principal
[[2, 11, 5]]
[[ 0  0  2 11  5]]
1/1 [==============================] - 0s 31ms/step
7
Going Marry: Sanghavi Sir is vice principle of SNJB's KBJ College of Engineering Chandwad
You:who is vice principal of 
who is vice principal of 
[[11, 5]]
[[ 0  0  0 11  5]]
1/1 [==============================] - 0s 27ms/step
4
Going Marry: There are total six deparments in our college including Computer , Mechanical, Civil , E & TC , AIDS , MBA etc.
You:number of department in collge
number of department in collge
[[18, 3]]
[[ 0  0  0 18  3]]
1/1 [==============================] - 0s 29ms/step
4
Going Marry: There are total six deparments in our college including Computer , Mechanical, Civil , E & TC , AIDS , MB